query/src/main/java/com/connexta/replication/query/QueryService.java

```diff
@@ -0,0 +1,219 @@
+   *
+   * @param site the new site configuration
+   */
+  public void update(Site site) {
+    synchronized (this) {
+      this.site = site;
+      try {
+        this.adapter = queryServiceTools.getAdapterFor(site);
+      } catch (AdapterException e) {
+        LOGGER.info(
+            "Failed to create an adapter for {} stopping the query service for now."
+                + site.getUrl());
+        stop();
+      }
+    }
+  }
+
+  /** Starts this query service. */
+  public void start() {
+    if (future == null || future.isDone()) {
+      future =
+          executor.scheduleAtFixedRate(
+              this::query, 0, queryServiceTools.getPeriod(), TimeUnit.SECONDS);
+    }
+  }
+
+  /** Stops this query service. */
+  public void stop() {
+    executor.shutdownNow();
+  }
+
+  @VisibleForTesting
+  void query() {
+    if (!adapter.isAvailable()) {
+      LOGGER.debug("System at " + site.getUrl() + " is currently unavailable");
+      return;
+    }
+
+    Set<Filter> filters = queryServiceTools.activeFiltersFor(site).collect(Collectors.toSet());
+    for (Filter filter : filters) {
+      FilterIndex filterIndex = queryServiceTools.getOrCreateFilterIndex(filter);
+
+      // holds the modified time of each task put in the queue for this filter
+      // this will be used to grab the latest modified time after
+      Set<Instant> modifiedTimes = new HashSet<>();
+
+      try {
+        getNewMetadataForFilter(filter, filterIndex)
+            .map(metadatum -> metadataToTask(metadatum, filter.getPriority()))
+            .forEach(task -> queueTaskAndPutLastModifiedInSet(task, modifiedTimes));
+      } catch (AdapterException e) {
+        LOGGER.debug(
+            "Couldn't query site {}. Stopping queries for this site until next polling attempt.",
+            site.getName(),
+            e);
+        return;
+      } catch (QueueException qe) {
+        LOGGER.debug(
+            "Failed to put a task in queue for site {}. Stopping queuing of tasks for this site. Will resume at next polling attempt.",
+            site.getName(),
+            qe);
+        return;
+      }
+      saveLastModifiedInFilterIndex(filterIndex, modifiedTimes);
+    }
+  }
+
+  private Stream<Metadata> getNewMetadataForFilter(Filter filter, FilterIndex filterIndex)
+      throws AdapterException {
+    // get last modified date
+    final Date indexDate = filterIndex.getModifiedSince().map(Date::from).orElse(null);
+    if (indexDate == null) {
+      LOGGER.trace("No index for filter {} found.", filter.getName());
+    }
+
+    // not implementing failure retries here
+    QueryRequest queryRequest =
+        new QueryRequestImpl(
+            filter.getFilter(),
+            Collections.emptyList(), // ignoring excluded nodes (origins tag)
+            Collections.emptyList(), // ignoring failed items
+            indexDate);
+
+    Iterable<Metadata> changeSet = adapter.query(queryRequest).getMetadata();
```

Comment:
❗️ ❓ How does the adapter deals with `InterrutedIOException`? Does it blindly convert it into an `AdapterException`? If so then we need to update the adapter to handle it differently as you need to be made aware that you got interrupted other than by receiving a generic error since being interrupted is not an error.